### PR TITLE
ApplicationTemplate Proposal

### PR DESCRIPTION
**Pull request by aszecowka**
In this proposal, following concepts are discussed:
- Integration System
- Application Template

Be aware that in the issue description (#339), there is a notion of OneTimeTokenService. It is not mentioned in this proposal, because IMO OneTimeTokenService is an implementation detail of IntegrationService and it should not be exposed in the Compass API. Such details should be raised in the `ApplicationClass.documentation`

### CODE REVIEW

File: docs/architecture/app-class-and-template.md [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+```
+Removing Integration System can be performed only when:
+- there is no Application Class from this Integration System
+- there is no Application managed by the Integration System 
+
+   
+2. Managing Application Class
+
+```
+input ApplicationClassInput {
```
Comment (crabtree) [inline docs/architecture/app-class-and-template.md]: Do we assume that on this level (ApplicationClass) the API specs are defined? Not sure if this is still valid, but there was a scenario when a particular instance of the Application can provide only a subset of its API for given spec credentials.

Comment (crabtree) [inline docs/architecture/app-class-and-template.md]: If we provide here the API and EventAPI specs on this level, shouldn't we provide also the documents to be consistent with the `Application`

**Comment by crabtree** [comment]
Generally, I would consider removing the specs from the `ApplicationClass` type. Right now I'm a bit confused about why the specs are in `ApplicationClass` and `ApplicationTemplate`. If I want to create the application that is managed by `IntegrationSystem` I can do it by creating the `Application` with `ApplicationClass` specified. Also, if I want to create an `Application `with the specs predefined the I would create an application from a template.

File: docs/architecture/app-class-and-template.md [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    description: String
+    documentation: Document
+    integrationSystemID: String!
+    apiSpecs: [ApiSpecInput]
+    eventApiSpecs: [EventApiSpecInput]
+}
+
+input ApplicationInput {
+    ...
+    applicationClass: String
```
[the remark this answers is not in the document]
Comment (aszecowka) [reply docs/architecture/app-class-and-template.md]: IMO it will be be better to refer Application Class by human-readable name, not id. This impose of course that ApplicationClass.Name is unique. So maybe I will change this field to applicationClassName?

File: docs/architecture/app-class-and-template.md [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+This kind of information is stored in the `ApplicationClass.document`.
+
+### Create Application Template
+1. Create Application Template
+```
+createApplicationTemplate(
+    in:{
+        name:"Y-Template", 
+        applicationInput:{
+              Name:"$APPLICATION_NAME",
```
[the remark this answers is not in the document]
Comment (aszecowka) [reply docs/architecture/app-class-and-template.md]: To be honest I don't want to decide it right now, for me both approaches are good, we should choose syntax hich will be easier for us to implement ;) maybe I will add a comment that exact syntax will be decided later.

File: docs/architecture/app-class-and-template.md [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+    })
+
+```
+2. Create Application from template
+```
+createApplicationFromTemplate(templateName: "Y-Template", values: [{Placeholder:"APPLICATION_NAME", Value:"MyApplication"}]): 
+
+```
+
+## Discussions
```
[the remark this answers is not in the document]
Comment (aszecowka) [reply docs/architecture/app-class-and-template.md]: renamed to FAQ

File: docs/architecture/app-class-and-template.md (hunk 1 above)
Comment (aszecowka) [reply docs/architecture/app-class-and-template.md]: In the first implementation, we make an assumption that ApplicationClass will be registered globally. Because of that, every tenant will see the same API. 
API is defined in `apiSpecs` and `eventAPISpecs` and they are just for documentation purposes. AS a user I would like to know what kind of API given ApplicationClass can provide. Notice that this is APISpecINput, no APIDefinitionInput so no credentials are defined at this level.

Comment (aszecowka) [reply docs/architecture/app-class-and-template.md]: I renamed `documentation` to `documents` which now is an array of Document objects

File: docs/architecture/app-class-and-template.md [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+## Terminology
+
+- **Application** represents functionality that provides API or Events.
+
+- **ApplicationClass** represents soon-to-be Application that is manged by IntegrationSystem. 
+   
+- **ApplicationTemplate** represents soon-to-be Application with all details already provided (like API and Events) that 
+can be created by providing values for replacing placeholders.
+  
+- **IntegrationSystem** (abbreviation: **IntSys**):
```
[the remark this answers is not in the document]
Comment (aszecowka) [reply docs/architecture/app-class-and-template.md]: I removed from terminology `Application` and `Integration System` and linked them from `terminology.md`. I keep `ApplicationClass` and `ApplicationTemplate` here - because these concepts are introduced in this document. In addition to that, I copied those definitions to `terminology.md`.

File: docs/architecture/app-class-and-template.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (aszecowka) [reply docs/architecture/app-class-and-template.md]: I added following sentence:
```In the example above, we use the following syntax `$VARIABLE_NAME` for defining placeholders. However, this is only an example 
and a final decision about syntax is postponed to the implementation phase. 
```

File: docs/architecture/app-class-and-template.md [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+or EventAPI provided by Applications originated from the ApplicationClass.
+In the first iteration, ApplicationClass can be registered only globally (`ApplicationClass.global`), so it will be available for every tenant. 
+To create the Application of given class, user set field `applicationClassName` in `ApplicationInput`. This field is optional.
+`deleteApplicationClass` is possible only if there are no classes originated from this ApplicationClass. 
+
+3. Managing Application Template 
+
+```
+input ApplicationTemplateInput {
+    name:               String!
```
Comment (PK85) [inline docs/architecture/app-class-and-template.md]: Please add description as well.

File: docs/architecture/app-class-and-template.md [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+    applicationInput:   ApplicationInput!
+    placeholders:       [PlaceholderDefinitionInput]
```
Comment (PK85) [inline docs/architecture/app-class-and-template.md]: Right now not in the scope. Nice feature for future. Add small section about feature called: "Placehiolders"

Comment (aszecowka) [reply docs/architecture/app-class-and-template.md]: I added a section called `Application Template Placeholders Implementation` with an explanation of why placeholder implementation seems to be easy. Why am I pushing toward this approach? 
First of all, in my opinion, it does not make sense to provide templates without possibility of customizing it - it will cause that we can create only one application from given template in the given tenant. 

We also considered an option about adding a new query that returns applicationInput for the given template.
It has at least 2 drawbacks:
- ApplictionInput cannot be used as a return value (Graphql limitation), because of that we have to introduce a counterpart for it. And that causes code duplication. What is even worse an ApplicationInput has many other inputs and they also have to be duplicated. 
- using this API can be cumbersome. At the beginning, you have to query for all fields from 
from ApplicationInput (this is how Graphql works), the response will be formatted in JSON format and then you have to manually convert it to Graphql again, here you can find a simplified example:

query:
```
  inputFromTemplate{
   Name
    Description
  }
}
```
response:
```
{
  "data": {
    "inputFromTemplate": {
      "Name": "name",
      "Description": "desc"
    }
  }
}
```
and mutation:
```
mutation {
  createApp(sth:{Name:"name", Description:"desc"}) {
    Name
    Description
  }
}
```

File: docs/architecture/app-class-and-template.md [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+   
+2. Managing Application Class
+
+```
+input ApplicationClassInput {
+    name: String!
+    description: String
+    documentation: Document
+    integrationSystemID: String!
+    apiSpecs: [ApiSpecInput]
```
[the remark this answers is not in the document]
Comment (aszecowka) [reply docs/architecture/app-class-and-template.md]: this field is renamed to previewApiSpec and it is optional

File: docs/architecture/app-class-and-template.md [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+    eventApiSpecs: [EventApiSpecInput]
```
[the remark this answers is not in the document]
Comment (aszecowka) [reply docs/architecture/app-class-and-template.md]: same as above

File: docs/architecture/app-class-and-template.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (aszecowka) [reply docs/architecture/app-class-and-template.md]: removed

File: docs/architecture/app-class-and-template.md [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+Because of that, we are going to extend our API with IntegrationSystem and ApplicationClass types.
+
+To increase the reusability of the application configuration, we introduce ApplicationTemplate. 
+
+
+## Terminology
+
+- **Application** represents functionality that provides API or Events.
+
+- **ApplicationClass** represents soon-to-be Application that is manged by IntegrationSystem.
```
[the remark this answers is not in the document]
Comment (aszecowka) [reply docs/architecture/app-class-and-template.md]: - added information about labelling

File: docs/architecture/app-class-and-template.md [hunk 11]
```diff
@@ -0,0 +1,200 @@
+# ApplicationClass and ApplicationTemplate
+
+Managing [Applications](./../terminology.md#Application) can be passed on some external system, to remove the burden of integration with Compass from Applications.
+Because of that, we are going to extend our API with [IntegrationSystem](./../terminology.md#Integration-System) and ApplicationClass types.
+
+To increase the reusability of the application configuration, we introduce ApplicationTemplate. 
+
+
+## Terminology
+- **ApplicationClass** represents soon-to-be Application that is managed by IntegrationSystem.
```
[the remark this answers is not in the document]
Comment (aszecowka) [reply docs/architecture/app-class-and-template.md]: done

File: docs/architecture/app-class-and-template.md [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+}
+
+```
+ApplicationTemplate can be treated as a prepopulated ApplicationInput with additional placeholders that have to be replaced by the user who creates Application from a template. 
+When defining ApplicationTemplate, all placeholders have to be documented in 
+`placeholders` field. To create an application from the template, a new mutation `createApplicationFromTemplate` is used.
+
+   
+## Workflows
+![](./assets/integration-system-and-app-class.svg)
```
[the remark this answers is not in the document]
Comment (aszecowka) [reply docs/architecture/app-class-and-template.md]: done

File: docs/architecture/app-class-and-template.md [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+### Create Application Class and Application
+![](./assets/app-class.svg)
+
+1. Integration System fetches access token from Hydra public API and then register Application Class using mutation
+ 
+```
+createApplicationClass(in:{name: "ClassX",integrationSystemID:"{ID}" })
+```
+
+1. User request for creating an application using Application Class with the following mutation:
```
Comment (crabtree) [inline docs/architecture/app-class-and-template.md]:
```suggestion
2. User request for creating an application using Application Class with the following mutation:
```

File: docs/architecture/app-class-and-template.md [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+
+To increase the reusability of the application configuration, we introduce ApplicationTemplate.
+The diagram below describes when to use ApplicationClass or ApplicationTemplate.
+
+![](./assets/app-vs-template-vs-class.svg)
+
+## Terminology
+
+- **ApplicationClass** represents an object for creating Applications. When you create an application from it, the IntegrationSystem
+dynamically populates definition.
```
Comment (pkosiec) [inline docs/architecture/app-class-and-template.md]: populates -> updates?
and you can mention here controlling some fields, such as API or EventAPI definitions

File: docs/architecture/app-class-and-template.md [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+`previewApiSpecs` and `previewEventApiSpecs` are optional fields that document what kind of API a user can expect after creating 
+Application originated from the ApplicationClass. The final API can differ from those presented in preview fields, for example, some API can be missing
+because the user lacks privileges to access them. 
+
+In the first iteration, ApplicationClass can be registered only globally (`ApplicationClass.global`), so it will be available for every tenant. 
+To create the Application of given class, the user set field `applicationClassName` in `ApplicationInput`. This field is optional.
+In case `applicationClassName` is provided, Compass adds two labels to the Application:
+- `applicationClass`
+- `inegrationSystemID`
+
```
Comment (pkosiec) [inline docs/architecture/app-class-and-template.md]: Also, please mention that some fields provided by user may be overwritten  by Integration System. For example, if user provides some APIs for Application managed by Integration System, Integration System should update the list with correct APIs. I'm missing explicit piece of information that some Application fields are controlled by Integration System and any user input will be overwritten by Integration System, which has to watch for changes and reset user changes for App.

File: docs/architecture/app-class-and-template.md [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+`placeholders` field. To create an application from the template, a new mutation `createApplicationFromTemplate` is used.
+
+   
+## Workflows
+### Create Application Class and Application
+![](./assets/app-class.svg)
+
+1. Integration System fetches access token from Hydra public API and then register Application Class using mutation
+ 
+```
```
Comment (pkosiec) [inline docs/architecture/app-class-and-template.md]:
```suggestion
```graphql
```

File: docs/architecture/app-class-and-template.md [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+createApplicationClass(in:{name: "ClassX",integrationSystemID:"{ID}" })
+```
+
+1. User request for creating an application using Application Class with the following mutation:
+```
```
Comment (pkosiec) [inline docs/architecture/app-class-and-template.md]:
```suggestion
```graphql
```

File: docs/architecture/app-class-and-template.md [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+         {
+            Name:"APPLICATION_NAME",
+            Description:"Name of the application"
+         }],
+    })
+
+```
+
+2. Create Application from template
+```
```
Comment (pkosiec) [inline docs/architecture/app-class-and-template.md]:
```suggestion
```graphql
```

File: docs/architecture/app-class-and-template.md [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+`deleteApplicationClass` is possible only if there are no classes originated from this ApplicationClass. 
+
+Extending API with ApplicationClass introduces new requirements for Integration Systems.
+Integration System can be removed only when:
+- there is no Application Class from this Integration System
+- there is no Application managed by the Integration System 
+
+2. Managing Application Template 
+
+```
```
Comment (pkosiec) [inline docs/architecture/app-class-and-template.md]:
```suggestion
```graphql
```

File: docs/architecture/app-class-and-template.md [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+- **ApplicationClass** represents an object for creating Applications. When you create an application from it, the IntegrationSystem
+dynamically populates definition.
+   
+- **ApplicationTemplate** represents an object for creating Applications from the static definition.
+
+## GraphQL API Changes
+   
+1. Managing Application Class
+
+```
```
Comment (pkosiec) [inline docs/architecture/app-class-and-template.md]:
```suggestion
```graphql
```

File: docs/architecture/app-class-and-template.md (hunk 15 above)
Comment (aszecowka) [reply docs/architecture/app-class-and-template.md]: I rephrased ApplicationClass definition:
`When you create an application from it, the IntegrationSystem dynamically updates Application definition, such us APIs, EventAPIs etc.`
>  I'm missing explicit piece of information that some Application fields are controlled by Integration System and any user input will be overwritten by Integration System, which has to watch for changes and reset user changes for App.

Currently, we don't have a clear concept about that, so at the beginning probably IntegrationSystem can update all fields.  Also, the behavior of IntegrationSystem is undefined if there are some APIs already registered manually.

File: docs/architecture/app-class-and-template.md [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+`ApplicationClass.documents` is a place for documenting any manual steps required for managing Application of given class by 
+a Integration System defined by `integrationSystemID`. For example, it can contain information on how to register a given application
+in the integration system if it is not performed by the Integration System automatically. 
+
+`previewApiSpecs` and `previewEventApiSpecs` are optional fields that document what kind of API a user can expect after creating 
+Application originated from the ApplicationClass. The final API can differ from those presented in preview fields, for example, some API can be missing
+because the user lacks privileges to access them. 
+
+In the first iteration, ApplicationClass can be registered only globally (`ApplicationClass.global`), so it will be available for every tenant.
```
[the remark this answers is not in the document]
Comment (aszecowka) [reply docs/architecture/app-class-and-template.md]: IMO it is better to not add such information because it is not yet discussed

File: docs/architecture/app-management.md [hunk 22]
```diff
@@ -0,0 +1,307 @@
...
+Managing [Applications](./../terminology.md#Application) can be passed on some external system, to remove from Applications the burden of integration with Compass.
+Because of that [IntegrationSystems](./components.md#Integration-System) were introduced.
+
+To simplify creation of Applications, Director API is extended with Application Template. 
+An ApplicationTemplate is reusable input for creating Applications that can be customized with placeholders provided 
+during Application creation.
+
+## Manage Applications by Integration System
+Managing Integrations System is described in [a separate document](./integration-systems.md).
+Integration System can be uniquely identified by it's name. To create an Application that is managed by a Integration System,
```
Comment (crabtree) [inline docs/architecture/app-management.md]: As this sounds like constraint I would rephrase this.
```suggestion
Integration System is uniquely identified by its name. To create an Application that is managed by an Integration System,
```

File: docs/architecture/app-management.md [hunk 23]
```diff
@@ -0,0 +1,307 @@
...
+}
+
+
+input PlaceholderDefinitionInput {
+    name            String!
+    description     String
+}
+
+input TemplateValueInput {
+    Placeholder     String!
```
Comment (crabtree) [inline docs/architecture/app-management.md]:
```suggestion
    placeholder     String!
```

File: docs/architecture/app-management.md [hunk 24]
```diff
@@ -0,0 +1,307 @@
...
+    Value           String!
```
Comment (crabtree) [inline docs/architecture/app-management.md]:
```suggestion
    value           String!
```

File: docs/architecture/app-management.md [hunk 25]
```diff
@@ -0,0 +1,307 @@
...
+            Name:"APPLICATION_NAME",
+            Description:"Name of the application"
+        }],
+        }) {
+           name
+    }
+ }
+```
+
+As you can see, `APPLICATION_NAME` placeholder is defined. In ApplicationInput, we refer to the placeholder in the following form: `{{APPLICATION_NAME}}`.
```
Comment (crabtree) [inline docs/architecture/app-management.md]: New line missing

File: docs/architecture/app-management.md [hunk 26]
```diff
@@ -0,0 +1,307 @@
...
+        integrationSystemName
+        labels
+    }
+}
+```
+Drawback of this approach is that label name (`simpleIntegrationSystem/applicationType`) and supported values(`ecommerce`,`marketing`) are arbitrary 
+defined by a IntegrationSystem and has to be documented. 
+Luckily, IntegrationSystem can simplify this process by defining ApplicationTemplates to explicitly specify supported types.
+
+## Managing ApplicationTemplates
```
Comment (crabtree) [inline docs/architecture/app-management.md]: I'm missing here a bit of background regarding the placeholders mechanism. Do we support only simple key-value pairs, or some function calls? Do we support some optional placeholders? Do we validate during the creation that all placeholders are filled with appropriate values?

File: docs/architecture/integration-systems.md [hunk 27]
```diff
@@ -0,0 +1,56 @@
...
+integration between Application and Compass is managed by the Integration System.
+
+## GraphQL API Changes
+
+GraphQL API is extended with mutations for managing Integration System, as well as 
+managing its Client Credentials.
+```
+type IntegrationSystem {
+    id: ID!
+    name: String!
```
Comment (crabtree) [inline docs/architecture/integration-systems.md]: Do we have some special requirements regarding the name? In the ApplicationCreateInput we refer to this name so there may be some rules that limit the naming.

File: docs/architecture/app-management.md [hunk 28]
```diff
@@ -0,0 +1,307 @@
...
+    webhooks: [WebhookInput!]
+    healthCheckURL: String
+    apis: [APIDefinitionInput!]
+    eventAPIs: [EventAPIDefinitionInput!]
+    documents: [DocumentInput!]
+    integrationSystemName: String #optional
+    labels: [LabelInput!]
+}
+```
+IntegrationSystemID is an optional. It means that you can still create Application that is not managed by IntegrationSystem.
```
Comment (dbadura) [inline docs/architecture/app-management.md]: Where is `IntegrationSystemID`. 
I see only `integrationSystemName`.

File: docs/architecture/app-management.md [hunk 29]
```diff
@@ -0,0 +1,315 @@
...
+Compass adds protected label with name `integrationSystemID` for just created Application, so output of the previous mutation is the following:
+```json
+{
+  "data": {
+    "createApplication": {
+      "id": "d046590f-934f-411f-91e2-d446b404a2a2",
+      "name": "simpleApplication",
+      "integrationSystemID": "aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa",
+      "labels": {
+        "scenarios":["DEFULT"],
```
Comment (crabtree) [inline docs/architecture/app-management.md]:
```suggestion
        "scenarios":["DEFAULT"],
```

File: docs/architecture/app-management.md [hunk 30]
```diff
@@ -0,0 +1,315 @@
+# Automatic Application Management
+
+Managing [Applications](./../terminology.md#Application) can be passed on some external system, to remove from Applications the burden of integration with Compass.
+Because of that [IntegrationSystems](./components.md#Integration-System) were introduced.
```
Comment (pkosiec) [inline docs/architecture/app-management.md]: Integration System

File: docs/architecture/app-management.md [hunk 31]
```diff
@@ -0,0 +1,315 @@
+# Automatic Application Management
```
Comment (pkosiec) [inline docs/architecture/app-management.md]: hmm, maybe it would be better to change file name - `app-management.md` is a bit too generic. What about `auto-app-management.md` or something like that?
